### PR TITLE
Move audio files over to S3

### PR DESCRIPTION
**Pull request by mlissner**
This is a fairly straightforward PR, I think. The idea is that on Friday we want to move our audio infrastructure over to S3. To do this, we have to change the storage class used in the model, migrate it, and tweak any code that is not implemented in the AWS storage class. That'll get most things up and running again. 

A few other pieces:

1. Nginx. As we migrate things to S3, we need redirection for our old content. I tweaked nginx accordingly, to send any audio-related visitors over to storage.cl.com.

2. Typing. I tried to make typing more complete.

3. Refactoring. I refactored our storage classes to remove duplication, and I refactored our tests to import data faster, and to allow skipping OCR to make them faster.

4. Finally, tests remain slightly broken, but they pass locally. Now that we have tests that rely on files in S3, we had the choice of either mocking the hell out of our tests so that they don't actually use S3 (which is doable), or putting test assets into S3 where the tests can find them. Doing all the mocks seemed like a bad solution, because (a), it was a lot of work, and (b), it meant we weren't testing things properly! The easiest way to mock it is to swap out the S3 storage engine for the default FileSystemStorage we use presently, but, well, now we're not testing things. 

    So, instead of doing the mock approach, I put copies of our test assets into our dev bucket. This means that tests hit the network (I think they already did?), and that we need to give our AWS dev creds over to Github via some secret. I haven't figured out that piece yet, but I will tomorrow.

@flooie, if you can give this a look over as soon as possible, I'd greatly appreciate it.

Fixes: #1033

### CODE REVIEW

File: cl/alerts/templates/alert_email.html [hunk 1]
```diff
@@ -47,7 +47,8 @@ <h3 class="alt bottom" style="font-size: 1.5em; font-weight: normal; line-height
                         From the court
                     </a>
                     &nbsp;&nbsp;|&nbsp;&nbsp;
-                    <a href="https://www.courtlistener.com/{{result.local_path}}" style="font-size: 100%; font-weight: inherit; font-family: inherit; color: #009; border: 0; font-style: inherit; padding: 0; text-decoration: none; vertical-align: baseline; margin: 0;">
+                    {# Provide link to S3 for OA for now, for everything later. #}
+                    <a href="https://{% if type == "oa" %}storage{% else %}www{% endif %}.courtlistener.com/{{result.local_path}}" style="font-size: 100%; font-weight: inherit; font-family: inherit; color: #009; border: 0; font-style: inherit; padding: 0; text-decoration: none; vertical-align: baseline; margin: 0;">
```
Comment (flooie) [inline cl/alerts/templates/alert_email.html:51]: The compactness took me a minute to suss out.  I suppose the lack of spacing is important to URL generation but if it isn't I think there is a better way to write this.

File: cl/audio/feeds.py [hunk 2]
```diff
@@ -69,9 +69,8 @@ def item_extra_kwargs(self, item):
         }
 
     def item_enclosure_url(self, item):
-        return (
-            "https://www.courtlistener.com/%s" % get_item(item)["local_path"]
-        )
+        path = get_item(item)["local_path"]
```
Comment (flooie) [inline cl/audio/feeds.py:72]: love the f-strings

File: cl/audio/models.py [hunk 3]
```diff
@@ -89,15 +90,15 @@ class Audio(AbstractDateTimeModel):
...
         blank=True,
         db_index=True,
     )
     local_path_original_file = models.FileField(
         help_text="The location, relative to MEDIA_ROOT, on the CourtListener "
         "server, where the original file is stored",
         upload_to=make_upload_path,
-        storage=IncrementingFileSystemStorage(),
+        storage=AWSMediaStorage(),
         db_index=True,
```
Comment (flooie) [inline cl/audio/models.py:102]: Seems, either the help text should be updated, or the naming convention should change?  

In LASC we used the naming convention `filepath_s3`.  I suppose you could keep the naming convention and update the help text to reflect its not on the CL server.

File: cl/audio/models.py [hunk 4]
```diff
@@ -236,7 +243,7 @@ def as_search_dict(self):
                 "duration": self.duration,
                 "source": self.source,
                 "download_url": self.download_url,
-                "local_path": str(getattr(self, "local_path_mp3", None)),
+                "local_path": deepgetattr(self, "local_path_mp3.name", None),
```
Comment (flooie) [inline cl/audio/models.py:246]: This `deepgetattr` is cute - and something I would use but its totally a black box

File: cl/lib/storage.py [hunk 5]
```diff
@@ -1,45 +1,61 @@
...
     We are only able to use it in places where users are not uploading files,
-    and we are instead creating them programatically (for example, via a
+    and we are instead creating them programmatically (for example, via a
     scraper).
 
     For more detail, see:
 
     https://docs.djangoproject.com/en/1.8/releases/1.5.9/#file-upload-denial-of-service
+
+    :param instance: The instance of the storage class being used
```
Comment (flooie) [inline cl/lib/storage.py:34]: This is very minor, but I would either use punctuation on all params - or none

File: cl/lib/storage.py [hunk 6]
```diff
@@ -50,39 +66,27 @@ class UUIDFileSystemStorage(FileSystemStorage):
...
-            name = os.path.join(
-                dir_name, "%s_%s%s" % (file_root, next(count), file_ext)
-            )
-        return name
+    def get_available_name(
+        self,
+        name: str,
+        max_length: Optional[int] = None,
+    ) -> str:
+        return get_name_by_incrementing(self, name, max_length)
```
Comment (flooie) [inline cl/lib/storage.py:92]: Why did you remove the docstrings?

File: cl/scrapers/management/commands/cl_scrape_opinions.py [hunk 7]
```diff
@@ -209,7 +209,7 @@ def add_arguments(self, parser):
             help="Disable duplicate aborting.",
         )
 
-    def scrape_court(self, site, full_crawl=False):
+    def scrape_court(self, site, full_crawl=False, ocr_available=True):
```
Comment (flooie) [inline cl/scrapers/management/commands/cl_scrape_opinions.py:212]: Missing a chance to add typing.

File: cl/audio/migrations/0010_switch_to_s3_storage_noop.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+    operations = [
+        migrations.AlterField(
+            model_name='audio',
+            name='local_path_mp3',
+            field=models.FileField(blank=True, db_index=True, help_text='The location, relative to MEDIA_ROOT, on the CourtListener server, where encoded file is stored', storage=cl.lib.storage.AWSMediaStorage(), upload_to=cl.lib.model_helpers.make_upload_path),
+        ),
+        migrations.AlterField(
+            model_name='audio',
+            name='local_path_original_file',
```
Comment (flooie) [inline cl/audio/migrations/0010_switch_to_s3_storage_noop.py:23]: Future comment, I think we should either alter the help text or update the fields to reflect the audio being on S3 and not the local courtlistener server?  We refer to files in LASC as s3_filepath.

File: cl/audio/templates/oral_argument.html [hunk 9]
```diff
@@ -216,12 +216,12 @@ <h3>{{ af.docket.court }}</h3>
...
                 {% elif not af.download_url and af.local_path_mp3 %}
-                    <a href="/{{ af.local_path_mp3 }}"
+                    <a href="{{ af.local_path_mp3.url }}"
                        class="btn btn-sm btn-primary">Download MP3</a>
                 {% elif af.download_url and af.local_path_mp3 %}
                     <div class="btn-group btn-group-sm" role="group"
                          aria-label="Download buttons">
-                        <a href="/{{ af.local_path_mp3 }}"
+                        <a href="{{ af.local_path_mp3.url }}"
                            class="btn btn-primary">Download MP3</a>
```
Comment (flooie) [inline cl/audio/templates/oral_argument.html:226]: this is much prettier.

File: docker/nginx/sites-available/courtlistener.com.conf [hunk 10]
```diff
@@ -18,13 +18,18 @@ server {
         expires max;
     }
     # White list of media files we serve directly
-    location ~ ^/(?:doc|docx|html|m4a|mp3|MP3|mp4|nsf|p|pdf|recap-thumbnails|rm|txt|wma|wpd|wsdl)/ {
```
Comment (flooie) [inline docker/nginx/sites-available/courtlistener.com.conf:21]: I'm not familiar with RM files --- is this correct?

File: mypy.ini [hunk 11]
```diff
@@ -15,6 +15,9 @@ ignore_missing_imports = True
 [mypy-juriscraper.*]
 ignore_missing_imports = True
 
+[mypy-localflavor.*]
```
Comment (flooie) [inline mypy.ini:18]: Im just going to assume you know what you are doing here.

File: .github/workflows/lint.yml [hunk 12]
```diff
@@ -73,8 +74,7 @@ jobs:
           cl/api/routers.py
           cl/api/tests.py
           cl/api/views.py
-          cl/opinion_page/views.py
```
Comment (flooie) [inline .github/workflows/lint.yml:76]: This is just general lack of knowledge on lint, but is removing opinion-page/views correct?

File: cl/audio/migrations/0010_switch_to_s3_storage_noop.sql [hunk 13]
```diff
@@ -0,0 +1,8 @@
+BEGIN;
+--
+-- Alter field local_path_mp3 on audio
+--
+--
+-- Alter field local_path_original_file on audio
```
Comment (flooie) [inline cl/audio/migrations/0010_switch_to_s3_storage_noop.sql:6]: See comment on naming.

File: cl/alerts/templates/alert_email.html (hunk 1 above)
Comment (mlissner) [reply cl/alerts/templates/alert_email.html:51]: This is only temporary, for two weeks, until we swap opinions alerts over to S3 too, so I think it's OK.

File: cl/audio/migrations/0010_switch_to_s3_storage_noop.py (hunk 8 above)
Comment (mlissner) [reply cl/audio/migrations/0010_switch_to_s3_storage_noop.py:23]: I decided against updating the fields themselves at least for now, because that'd break the API, but you're right, the help text needs a rework. Good one.

File: cl/lib/storage.py (hunk 5 above)
Comment (mlissner) [reply cl/lib/storage.py:34]: Good call. I'll clean it up.

File: cl/lib/storage.py (hunk 6 above)
Comment (mlissner) [reply cl/lib/storage.py:92]: I think it'd be mostly redundant here, since it calls out to a single function that has a long docstring.

File: cl/scrapers/management/commands/cl_scrape_opinions.py (hunk 7 above)
Comment (mlissner) [reply cl/scrapers/management/commands/cl_scrape_opinions.py:212]: I think I added typing here at first, but cut my losses when it didn't easily work.

File: docker/nginx/sites-available/courtlistener.com.conf (hunk 10 above)
Comment (mlissner) [reply docker/nginx/sites-available/courtlistener.com.conf:21]: Realmedia. I kid you not.

File: mypy.ini (hunk 11 above)
Comment (mlissner) [reply mypy.ini:18]: This is what you have to do for each library that lacks type hints to tell mypy to ignore it.

File: .github/workflows/lint.yml (hunk 12 above)
Comment (mlissner) [reply .github/workflows/lint.yml:76]: Oh weird. Good catch. Fixed.
